sdk/src/main/java/com/deploygate/sdk/CustomLogInstructionSerializer.java

```diff
@@ -0,0 +1,368 @@
+            }
+
+            return SEND_LOG_RESULT_FAILURE_RETRIABLE;
+        }
+    }
+
+    private void ensureHandlerInitialized() {
+        if (handler != null) {
+            return;
+        }
+
+        synchronized (configuration) {
+            if (handler != null) {
+                return;
+            }
+
+            handler = new CustomLogHandler(thread.getLooper(), this, configuration.backpressure, configuration.bufferSize);
+        }
+    }
+
+    /**
+     * This handler behaves as a ordered-buffer of instructions.
+     * <p>
+     * The instruction of adding a new log and sending buffered logs are synchronized.
+     */
+    private static class CustomLogHandler extends Handler {
+        private static final int WHAT_SEND_LOGS = 0x30;
+        private static final int WHAT_ADD_NEW_LOG = 0x100;
+
+        private final CustomLogInstructionSerializer serializer;
+        private final CustomLogConfiguration.Backpressure backpressure;
+        private final int bufferSize;
+        private final int maxWhatOffset;
+        private final LinkedList<CustomLog> customLogs;
+        private int pushWhatOffset = 0;
+
+        /**
+         * @param looper
+         *         Do not use Main Looper to avoid wasting the main thread resource.
+         * @param serializer
+         *         an instance to send logs
+         * @param backpressure
+         *         the backpressure strategy of the log buffer, not of instructions.
+         * @param bufferSize
+         *         the max size of the log buffer, not of instructions.
+         */
+        CustomLogHandler(
+                Looper looper,
+                CustomLogInstructionSerializer serializer,
+                CustomLogConfiguration.Backpressure backpressure,
+                int bufferSize
+        ) {
+            super(looper);
+            this.serializer = serializer;
+            this.backpressure = backpressure;
+            this.bufferSize = bufferSize;
+            this.maxWhatOffset = bufferSize * 2;
+            this.customLogs = new LinkedList<>();
+        }
+
+        /**
+         * Cancel the send-logs instruction in the handler message queue.
+         * This doesn't interrupt the thread and stop the sending-logs instruction that is running at the time.
+         */
+        void cancelPendingSendLogsInstruction() {
+            removeMessages(WHAT_SEND_LOGS);
+        }
+
+        /**
+         * Enqueue the send-logs instruction in the handler message queue unless enqueued.
+         */
+        void enqueueSendLogsInstruction() {
+            if (hasMessages(WHAT_SEND_LOGS)) {
+                return;
+            }
+
+            sendEmptyMessage(WHAT_SEND_LOGS);
+        }
+
+        /**
+         * Enqueue new add-new-log instruction to the handler message queue.
+         */
+        void enqueueAddNewLogInstruction(CustomLog log) {
+            Message msg = obtainMessage(WHAT_ADD_NEW_LOG + getAndIncrementPushWhatOffset(), log);
```

Comment:
In the Handler queue `what` can be duplicated so you don't need to increment it for each log, I guess? (`obtainMessage(what, obj)` just obtains a new `Message` instance and set `what`)